code/datums/components/harvester.dm

```diff
@@ -0,0 +1,142 @@
+
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine)
+	RegisterSignal(parent, COMSIG_ITEM_UNIQUE_ACTION, .proc/unique_action)
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK, .proc/attack)
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/attackby)
+
+/datum/component/chem_booster/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(
+		COMSIG_PARENT_EXAMINE,
+		COMSIG_ITEM_UNIQUE_ACTION,
+		COMSIG_ITEM_ATTACK,
+		COMSIG_PARENT_ATTACKBY))
+
+/datum/component/harvester/proc/examine(datum/source, mob/user)
+	SIGNAL_HANDLER
+	to_chat(user, span_rose("[length(beaker.reagents.reagent_list) ? "It currently holds [beaker.reagents.total_volume]u of [beaker.reagents.reagent_list[1].name]" : "The internal storage is empty"].\n<b>Compatible chemicals:</b>"))
+	for(var/R in loadable_reagents)
+		var/atom/L = R
+		to_chat(user, "[initial(L.name)]")
+
+/datum/component/harvester/proc/attackby(datum/source, obj/item/I, mob/user)
+	if(user.do_actions)
+		return TRUE
+
+	if(!isreagentcontainer(I) || istype(I, /obj/item/reagent_containers/pill))
+		to_chat(user, span_rose("[I] isn't compatible with [source]."))
+		return TRUE
+
+	var/trans
+	var/obj/item/reagent_containers/container = I
+
+	if(!container.reagents.total_volume)
+		trans = beaker.reagents.trans_to(container, 30)
+		to_chat(user, span_rose("[trans ? "You take [trans]u out of the internal storage. It now contains [beaker.reagents.total_volume]u" : "[source]'s storage is empty."]."))
+		return TRUE
+
+	if(length(container.reagents.reagent_list) > 1)
+		to_chat(user, span_rose("The solution needs to be uniform and contain only a single type of reagent to be compatible."))
+		return TRUE
+
+	if(beaker.reagents.total_volume && (container.reagents.reagent_list[1].type != beaker.reagents.reagent_list[1].type))
+		to_chat(user, span_rose("[source]'s internal storage can contain only one kind of solution at the same time. It currently contains <b>[beaker.reagents.reagent_list[1].name]</b>"))
+		return TRUE
+
+	if(!locate(container.reagents.reagent_list[1].type) in loadable_reagents)
+		to_chat(user, span_rose("This reagent is not compatible with the weapon's mechanism. Check the engraved symbols for further information."))
+		return TRUE
+
+	if(container.reagents.total_volume < 5)
+		to_chat(user, span_rose("At least 5u of the substance is needed."))
+		return TRUE
+
+	if(beaker.reagents.total_volume >= 30)
+		to_chat(user, span_rose("The internal storage is full."))
+		return TRUE
+
+	to_chat(user, span_notice("You begin filling up the [source] with [container.reagents.reagent_list[1]]."))
+	if(!do_after(user, 1 SECONDS, TRUE, source, BUSY_ICON_BAR, null, PROGRESS_BRASS))
+		return TRUE
+
+	trans = container.reagents.trans_to(beaker, container.amount_per_transfer_from_this)
+	to_chat(user, span_rose("You load [trans]u into the internal system. It now holds [beaker.reagents.total_volume]u."))
+	return TRUE
+
+/datum/component/harvester/proc/unique_action(datum/source, mob/user)
```

Comment:
every /proc/ must be autodoced, aka :
///Signal handler calling the unique action of the harvester